lib/utils/hostname.go

```diff
@@ -0,0 +1,104 @@
+	}
+	return fmt.Sprintf("*.%s", strings.Join(strings.Split(hostname, ".")[1:], "."))
+}
+
+// AddHostToHostList adds the provided hostname to the hostList, then processes the hostList to add wildcards
+// if appropriate. it returns a processed, sorted hostList.
+//
+// for every host in the list:
+//   - all hostnames are converted to their wildcard form (test.example.com -> *.example.com) and grouped together.
+//   - if we only have a single hostname which would be matched by a given wildcard, that hostname is added
+//     to the hostList verbatim to prevent matching against too many hosts with a wildcard.
+//   - if a wildcard matches multiple hostnames, the wildcard will be added to the hostList instead.
+//
+// this is an effort to keep the length of hostList as short as possible for efficiency and tidiness.
+func AddHostToHostList(hostList []string, hostname string) []string {
+	// add the incoming hostname to the hostList before we sort and process it
+	hostList = append(hostList, hostname)
+
+	hostMap := make(map[string][]string)
+	// iterate over the full hostList
+	// if the element is a wildcard, add it to the list of wildcards
+	// if the element is not a wildcard, convert it to a wildcard and add any hostnames it matches to a map
+	for _, element := range hostList {
+		// FQDN-based hosts are grouped under a wildcard key
+		if isFQDN(element) {
+			wildcard := wildcardFromHostname(element)
+			if !slices.Contains(hostMap[wildcard], element) {
+				hostMap[wildcard] = append(hostMap[wildcard], element)
+			}
+		} else {
+			// any non-wildcard hosts go into the "extra" key and will be processed separately at the end
+			hostMap["extra"] = append(hostMap["extra"], element)
+		}
+	}
+
+	// iterate over the map, look for all wildcard keys with more than one hostname matching.
+	// for each match, add the wildcard to the hostList.
+	var outputHostList []string
+	for key, matchingHostnames := range hostMap {
+		// add all non-wildcard matches separately
+		if key == "extra" {
+			for _, hostname := range matchingHostnames {
+				if !slices.Contains(outputHostList, hostname) {
+					outputHostList = append(outputHostList, hostname)
+				}
+			}
+		} else {
+			// add all wildcards with more than one hostname matching
+			if len(matchingHostnames) > 1 {
+				outputHostList = append(outputHostList, key)
+			} else {
+				// add the single hostname which is matched by the given wildcard
+				outputHostList = append(outputHostList, matchingHostnames[0])
+			}
+		}
+	}
+
+	slices.Sort(outputHostList)
+	return outputHostList
+}
+
+// NaivelyValidateHostname checks the provided hostname against a naive regex to ensure it doesn't contain obviously
+// illegal characters. It's not guaranteed to be perfect, just a simple sanity check.
+func NaivelyValidateHostname(hostname string) bool {
+	hostnameRegexp := regexp.MustCompile("^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]).)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])$")
```

Comment:
```suggestion

var hostnameRegexp = regexp.MustCompile("^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]).)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])$")


// NaivelyValidateHostname checks the provided hostname against a naive regex to ensure it doesn't contain obviously
// illegal characters. It's not guaranteed to be perfect, just a simple sanity check.
func NaivelyValidateHostname(hostname string) bool {
```

Move the compilation out so you don't have to recompile the regex every time.